src/Functions/ReplaceStringImpl.h

```diff
@@ -85,34 +96,275 @@ struct ReplaceStringImpl
+
+        ColumnString::Offset res_offset = 0;
+
+        size_t prev_haystack_offset = 0;
+        size_t prev_needle_offset = 0;
+        size_t prev_replacement_offset = 0;
+
+        for (size_t i = 0; i < haystack_size; ++i)
+        {
+            const auto * const cur_haystack_data = &haystack_data[prev_haystack_offset];
+            const size_t cur_haystack_length = haystack_offsets[i] - prev_haystack_offset - 1;
+
+            const auto * const cur_needle_data = &needle_data[prev_needle_offset];
+            const size_t cur_needle_length = needle_offsets[i] - prev_needle_offset - 1;
+
+            const auto * const cur_replacement_data = &replacement_data[prev_replacement_offset];
+            const size_t cur_replacement_length = replacement_offsets[i] - prev_replacement_offset - 1;
+
+            if (cur_needle_length == 0)
+                throw Exception(ErrorCodes::ARGUMENT_OUT_OF_BOUND, "Length of the pattern argument in function {} must be greater than 0.", name);
+
+            /// Using "slow" "stdlib searcher instead of Volnitsky because there is a different pattern in each row
+            StdLibASCIIStringSearcher</*CaseInsensitive*/ false> searcher(cur_needle_data, cur_needle_length);
+
+            const auto * last_match = static_cast<UInt8 *>(nullptr);
+            const auto * start_pos = cur_haystack_data;
+            const auto * const cur_haystack_end = cur_haystack_data + cur_haystack_length;
+
+            while (start_pos < cur_haystack_end)
+            {
+                if (const auto * const match = searcher.search(start_pos, cur_haystack_end); match != cur_haystack_end)
+                {
+                    /// Copy prefix before match
+                    res_data.resize(res_data.size() + (match - start_pos));
+                    memcpy(&res_data[res_offset], start_pos, match - start_pos);
+                    res_offset += (match - start_pos);
+
+                    /// Insert replacement for match
+                    res_data.resize(res_data.size() + cur_replacement_length);
+                    memcpy(&res_data[res_offset], cur_replacement_data, cur_replacement_length);
+                    res_offset += cur_replacement_length;
+
+                    last_match = match;
+                    start_pos = match + cur_needle_length;
+
+                    if constexpr (replace == ReplaceStringTraits::Replace::First)
+                        break;
+                }
+                else
+                    break;
+            }
+
+            /// Copy suffix after last match
+            size_t bytes = (last_match == nullptr) ? (cur_haystack_end - cur_haystack_data + 1)
+                                                   : (cur_haystack_end - last_match - cur_needle_length + 1);
+            res_data.resize(res_data.size() + bytes);
+            memcpy(&res_data[res_offset], start_pos, bytes);
+            res_offset += bytes;
+
+            res_offsets[i] = res_offset;
```

Comment:
Maybe this part of code can be unified between `vectorVectorConstant`, `vectorConstantVector` and `vectorVectorVector` like it's done in `RegexSearcher` with `processString` method?

Reply:
I'll give it a try

Reply:
I tried an approach similar to `processString`, but the code does not really get better by that. Reason is that there is a lot less shared code here than in `processString`. What I did was to factorize the copy code, that saves a little bit at least.